background.js

```diff
@@ -10,21 +21,159 @@ chrome.runtime.onMessage.addListener(async (message) => {
     }
   });
 
+// Checks if URL should be auto-archived based on regex patterns and configuration settings.
+async function shouldAutoArchive(url) {
+  try {
+    console.debug(`[Auto-Archive Debug] Checking URL: ${url}`);
+
+    const { enable_auto_archive=false, match_urls=[], exclude_urls=[] } = await chrome.storage.local.get([
+      'enable_auto_archive',
+      'match_urls',
+      'exclude_urls',
+    ]);
+
+    console.debug(`[Auto-Archive Debug] Settings: enable_auto_archive=${enable_auto_archive}, match_urls="${match_urls}", exclude_urls="${exclude_urls}"`);
+
+    if (!enable_auto_archive || !match_urls || match_urls.trim() === '') {
+      console.debug('[Auto-Archive Debug] Auto-archiving disabled or match pattern empty');
+      return false;
+    }
+
+    const matchPattern = new RegExp(match_urls);
+    const matches = matchPattern.test(url);
+    console.debug(`[Auto-Archive Debug] URL match test: ${matches} (pattern: ${matchPattern})`);
+
+    if (!matches) {
+      return false;
+    }
+
+    if (exclude_urls.trim()) {
+      try {
+        const excludePattern = new RegExp(exclude_urls);
+        const excluded = excludePattern.test(url);
+        console.debug(`[Auto-Archive Debug] URL exclude test: ${excluded} (pattern: ${excludePattern})`);
+
+        if (excluded) {
+          return false;
+        }
+      } catch (error) {
+        console.error('Invalid exclude pattern:', error);
+      }
+    }
+
+    console.debug(`[Auto-Archive Debug] URL ${url} should be archived: TRUE`);
+    return true;
+  } catch (error) {
+    console.error('Error checking auto-archive patterns:', error);
+    return false;
+  }
+}
+
+// Global reference to the listener so we can remove it properly
+let tabUpdateListener = null;
+
+async function setupAutoArchiving() {
+  console.debug('[Auto-Archive Debug] Setting up auto-archiving...');
+
+  const hasPermission = await chrome.permissions.contains({ permissions: ['tabs'] });
+  console.debug(`[Auto-Archive Debug] Has tabs permission: ${hasPermission}`);
+
+  if (!hasPermission) {
+    console.log('Tabs permission not granted, auto-archiving disabled');
+    // Reset the toggle if permission was not granted
+    chrome.storage.local.set({ enable_auto_archive: false });
+    return;
+  }
+
+  if (tabUpdateListener) {
+    try {
```

Comment:
why does this remove the listener here? can you add a comment explaining it

Reply:
Since we auto-archive by listening for updates to tabs, to disable auto-archiving we remove the tabUpdateListener. setupAutoArchiving runs every time the user toggles the auto-archiving checkbox, so removing the listener every time makes sure multiple listeners aren't registered.

We could also simplify the code, not mess with the listener, and just add the check for if auto-archiving is enabled directly in the tabUpdateListener, but then, if the feature is disabled, some unnecessary work would be done on every tab update.

I'll keep the same method for now and add some comments.